include/open62541/util.h

```diff
@@ -47,27 +54,84 @@ typedef enum {
  * invalidates pointers into the previous array. If the key exists already, the
  * value is overwritten. */
 UA_EXPORT UA_StatusCode
-UA_KeyValueMap_set(UA_KeyValuePair **map, size_t *mapSize,
+UA_KeyValueMap_set(UA_KeyValueMap *map,
                    const UA_QualifiedName key,
                    const UA_Variant *value);
+UA_EXPORT UA_StatusCode
+UA_KeyValueMap_setScalar(UA_KeyValueMap *map,
+                         const UA_QualifiedName key,
+                         void * UA_RESTRICT p,
+                         const UA_DataType *type);
 
 /* Returns a pointer to the value or NULL if the key is not found.*/
 UA_EXPORT const UA_Variant *
-UA_KeyValueMap_get(const UA_KeyValuePair *map, size_t mapSize,
+UA_KeyValueMap_get(const UA_KeyValueMap *map,
                    const UA_QualifiedName key);
 
+/* Wether or not the given key-value-map is empty.
+ * Returns true if map is NULL or map->mapSize is equal to zero,
+ * otherwise false.
+ * */
+UA_EXPORT UA_Boolean
+UA_KeyValueMap_isEmpty(const UA_KeyValueMap *map);
+
 /* Returns NULL if the value for the key is not defined or not of the right
  * datatype and scalar/array */
 UA_EXPORT const void *
-UA_KeyValueMap_getScalar(const UA_KeyValuePair *map, size_t mapSize,
+UA_KeyValueMap_getScalar(const UA_KeyValueMap *map,
                          const UA_QualifiedName key,
                          const UA_DataType *type);
 
 /* Remove a single entry. To delete the entire map, use UA_Array_delete. */
 UA_EXPORT UA_StatusCode
-UA_KeyValueMap_delete(UA_KeyValuePair **map, size_t *mapSize,
+UA_KeyValueMap_remove(UA_KeyValueMap *map,
                       const UA_QualifiedName key);
 
+UA_EXPORT UA_KeyValueMap *
+UA_KeyValueMap_new(size_t size_hint);
+
+UA_EXPORT void
+UA_KeyValueMap_delete(UA_KeyValueMap *map);
+
+/**
+ * Create a deep copy of the given KeyValueMap. Function will allocate memory for
+ * the newly create KeyValueMap which needs to be freed by the user of the function
+ * accordingly.
+ * @param map The KeyValueMap to be copeid
+ * @param mapSize The size of the KeyValueMap
+ * @return A pointer to the copied KeyValueMap, NULL if memory allocation failed.
+ */
+UA_EXPORT UA_KeyValueMap*
+UA_KeyValueMap_copy(UA_KeyValueMap *map);
```

Comment:
take two input pointers and return a statuscode.
The target map could be a stack-allocated struct.
Keep the API similar to `UA_copy` et al.
The input map can be const.